openvdb/openvdb/tools/Merge.h

```diff
@@ -0,0 +1,931 @@
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+
+/// @brief Convenience class that contains a pointer to a const or a non-const tree
+/// and a subset of methods to retrieve data from the tree.
+///
+/// @details This class has methods to extract data from the tree that will either
+/// result in stealing the data or deep-copying it depending on whether the
+/// DeepCopy or Steal tag dispatch class was used in its construction.
+///
+/// @details Stealing data requires a non-const tree pointer. There is a constructor
+/// to pass in a tree shared pointer for cases where it is desirable for this class
+/// to maintain shared ownership.
+template <typename TreeT>
+struct TreeToMerge
+{
+    using TreeType = std::remove_const_t<TreeT>;
+    using RootNodeType = typename TreeType::RootNodeType;
+    using ValueType = typename TreeType::ValueType;
+    using MaskTreeType = typename TreeT::template ValueConverter<ValueMask>::Type;
+
+    TreeToMerge() = default;
+
+    /// @brief Non-const pointer tree constructor for stealing data.
+    TreeToMerge(TreeType* tree, Steal)
+        : mTree(tree) { }
+    /// @brief Non-const shared pointer tree constructor for stealing data.
+    TreeToMerge(typename TreeType::Ptr treePtr, Steal)
+        : mTreePtr(treePtr), mTree(mTreePtr.get()) { }
+
+    /// @brief Const tree pointer constructor for deep-copying data. As the
+    /// tree is not mutable and thus cannot be pruned, a lightweight mask tree
+    /// with the same topology is created that can be pruned to use as a
+    /// reference. Initialization of this mask tree can optionally be disabled
+    /// for delayed construction.
+    TreeToMerge(const TreeType* constTree, DeepCopy, bool initialize = true)
+        : mConstTree(constTree)
+    {
+        if (mConstTree && initialize)     this->initializeMask();
+    }
+
+    /// @brief Non-const tree pointer constructor for deep-copying data. The
+    /// tree is not intended to be modified so is not pruned, instead a
+    /// lightweight mask tree with the same topology is created that can be
+    /// pruned to use as a reference. Initialization of this mask tree can
+    /// optionally be disabled for delayed construction.
+    TreeToMerge(TreeType* tree, DeepCopy tag, bool initialize = true)
+        : TreeToMerge(static_cast<const TreeType*>(tree), tag, initialize) { }
+
+    /// @brief Reset the non-const tree shared pointer. This is primarily
+    /// used to preserve the order of trees to merge in a container but have
+    /// the data in the tree be lazily loaded or resampled.
+    void reset(typename TreeType::Ptr treePtr)
+    {
+        mTreePtr = treePtr;
+        mTree = mTreePtr.get();
+        mConstTree = nullptr;
+    }
+
+    /// @brief Return a pointer to the non-const tree.
+    TreeType* tree() { return mTree; }
+    /// @brief Return a pointer to the const tree.
+    const TreeType* constTree() { return mConstTree; }
```

Comment:
These are a bit of a misnomer.
stealTree()
deepCopyTree()
would be more accurate as they reflect if it is in that mode or not.

Otherwise, it seems odd that constTree() doesn't return mTree if mConstTree is nullptr; as that is also a valid const tree!

Those sound like verbs though.  treeToStealFrom().  treeToDeepCopyFrom()?

Reply:
Yes, I agree with you, let's remove the const/non-const differentiation entirely and use steal/deep-copy. `stealTree()` is clearly confusing, perhaps we just drop the From - `TreeToMerge::treeToSteal()`?

Reply:
Yep, dropping the From sounds wise.